Firestore/core/src/firebase/firestore/api/query_core.mm

```diff
@@ -0,0 +1,327 @@
+        : firestore_(parent->firestore()),
+          query_(parent->query()),
+          user_listener_(std::move(user_listener)) {
+    }
+
+    void OnEvent(StatusOr<ViewSnapshot> maybe_snapshot) override {
+      if (!maybe_snapshot.status().ok()) {
+        user_listener_->OnEvent(maybe_snapshot.status());
+        return;
+      }
+
+      ViewSnapshot snapshot = std::move(maybe_snapshot).ValueOrDie();
+      SnapshotMetadata metadata(snapshot.has_pending_writes(),
+                                snapshot.from_cache());
+
+      QuerySnapshot result(firestore_, query_, std::move(snapshot),
+                           std::move(metadata));
+
+      user_listener_->OnEvent(result);
+    }
+
+   private:
+    std::shared_ptr<Firestore> firestore_;
+    FSTQuery* query_;
+    QuerySnapshot::Listener user_listener_;
+  };
+  auto view_listener =
+      absl::make_unique<Converter>(this, std::move(user_listener));
+
+  // Call the view_listener on the user Executor.
+  auto async_listener = AsyncEventListener<ViewSnapshot>::Create(
+      firestore_->client().userExecutor, std::move(view_listener));
+
+  std::shared_ptr<QueryListener> query_listener =
+      [firestore_->client() listenToQuery:this->query()
+                                  options:options
+                                 listener:async_listener];
+
+  return ListenerRegistration(firestore_->client(), std::move(async_listener),
+                              std::move(query_listener));
+}
+
+Query Query::Filter(FieldPath field_path,
+                    Filter::Operator op,
+                    FSTFieldValue* field_value,
+                    const std::function<std::string()>& type_describer) const {
+  if (field_path.IsKeyFieldPath()) {
+    if (op == Filter::Operator::ArrayContains) {
+      ThrowInvalidArgument(
+          "Invalid query. You can't perform arrayContains queries on document "
+          "ID since document IDs are not arrays.");
+    }
+    if (field_value.type == FieldValue::Type::String) {
+      const std::string& document_key =
+          static_cast<FSTDelegateValue*>(field_value)
+              .internalValue.string_value();
+      if (document_key.empty()) {
+        ThrowInvalidArgument(
+            "Invalid query. When querying by document ID you must provide a "
+            "valid document ID, but it was an empty string.");
+      }
+      if (![query() isCollectionGroupQuery] &&
+          document_key.find('/') != std::string::npos) {
+        ThrowInvalidArgument(
+            "Invalid query. When querying a collection by document ID you must "
+            "provide a plain document ID, but '%s' contains a '/' character.",
+            document_key);
+      }
+      ResourcePath path =
+          query().path.Append(ResourcePath::FromString(document_key));
+      if (!DocumentKey::IsDocumentKey(path)) {
+        ThrowInvalidArgument(
+            "Invalid query. When querying a collection group by document ID, "
```

Comment:
Question: why is it certain that this is a collection group query?

Reply:
I'm not entirely sure. I've double-checked that the logic is unmodified from the source. @mikelehen, could you comment on why this is certainly a collection group query at this point? That doesn't seem to be supported by the ifs above.

Reply:
It's certain because of the underlying knowledge that `query().path` will be even-length if-and-only-if it's a collection group query (because collection group queries are rooted on a document while collection queries are rooted on a collection).  So when we hit this error because `IsDocumentKey(path)` is not true (i.e. `path` is odd-length), there are two possibilities:
1. It's a collection query (odd-length) combined with an even-length document_key.
2. It's a collection group query (even-length) combined with an odd-length document_key.

Option 1 is covered by the prior if statement, since to have an even-length document_key it must contain slashes.  So if we have hit this error it's because of option 2 and that implies it's a collection group query.

If this is too non-obvious, feel free to log a bug against me and I'll go ahead and comment / rework to make it clearer.